src/operator/contrib/ngraph-inl.h

```diff
@@ -0,0 +1,148 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2018 Intel Corporation
+ * \file ngraph.cc
+ * \brief ngraph subgraph property for mxnet
+*/
+
+#ifndef MXNET_OPERATOR_CONTRIB_NGRAPH_INL_H_
+#define MXNET_OPERATOR_CONTRIB_NGRAPH_INL_H_
+
+#if MXNET_USE_NGRAPH
+#include <ngraph_compiler.h>
+#include <ngraph_imperative.h>
+#include <ngraph_nnvm_ops.h>
+#include <ngraph_utils.h>
+
+#include <vector>
+
+#include "../subgraph/common.h"
+#include "../subgraph/subgraph_property.h"
+
+namespace mxnet {
+namespace op {
+
+// when built with NGRAPH we use this subgraph by default
+static int ngraph_backend = setenv("MXNET_SUBGRAPH_BACKEND", "ngraph", 0);
+
+class SgNgraphSelector : public SubgraphSelector {
+ public:
+  explicit SgNgraphSelector(ngraph_bridge::Compiler *compiler)
+      : compiler_(compiler), valid(compiler_->get_node_map().size() > 0) {}
+
+  bool Select(const nnvm::Node &n) override { return is_node_selected(n); }
+
+  bool SelectInput(const nnvm::Node &n, const nnvm::Node &new_node) override {
+    return is_node_selected(n, &new_node);
+  }
+
+  bool SelectOutput(const nnvm::Node &n, const nnvm::Node &new_node) override {
+    return is_node_selected(n, &new_node);
+  }
+  std::vector<nnvm::Node *> Filter(
+      const std::vector<nnvm::Node *> &candidates) {
+    if (candidates.size() == 1 && candidates[0]->inputs.size() == 0) {
+      return std::vector<nnvm::Node *>();
+    } else {
+      return candidates;
+    }
+  }
+
+ private:
+  ngraph_bridge::Compiler *compiler_;
+  const bool valid;
+  ngraph_bridge::NodePtr get_node(const nnvm::Node *n) {
+    if (n) {
+      auto &entry_map = compiler_->get_ngraph().entry_map_;
+      ngraph_bridge::MapEntry tmp{compiler_->get_node_map().at(n).get(), 0};
+      if (entry_map.count(tmp)) {
+        return entry_map[tmp];
+      }
+    }
+    return nullptr;
+  }
+  bool is_node_selected(const nnvm::Node &n, const nnvm::Node *next = nullptr) {
+    bool selected = false;
+    if (!valid) return selected;
+
+    auto nn = get_node(&n);
+    auto nnext = get_node(next);
+
+    selected = nn && nn->in_ngraph_;
+    if (next) {
+      selected =
+          selected && nnext->in_ngraph_ && nn->subgraph_ == nnext->subgraph_;
+    }
+    return selected;
+  }
+};
+
+class SgNgraphProperty : public SubgraphProperty {
+ public:
+  static SubgraphPropertyPtr Create() {
+    return std::make_shared<SgNgraphProperty>();
+    if (ngraph_backend != 0 && ngraph_bridge::ngraph_log_verbose_detail) {
+      std::cout
+          << "NGRAPH_BRIDGE: WARNING, failed to set MXNET_SUBGRAPH_BACKEND"
+          << std::endl;
+    }
```

Comment:
what is the if condition for?

Reply:
Debugging, mostly, it's only enabled when a debugging environment variable is set.

Reply:
the function already returns before the if condition. I don't see how it can be used for debugging. If you want to print something, i think it's better to use LOG().

Reply:
...I'm not sure how that got through, probably a last minute refactor, will fix, thanks.